library/std/src/io/readbuf.rs

```diff
@@ -243,3 +251,135 @@ impl<'a> ReadBuf<'a> {
         self.initialized
     }
 }
+
+/// A wrapper around [`&mut ReadBuf`](ReadBuf) which prevents the buffer that the `ReadBuf` points to from being replaced.
+#[derive(Debug)]
+pub struct ReadBufRef<'a, 'b> {
+    read_buf: &'a mut ReadBuf<'b>,
+}
+
+impl<'a, 'b> ReadBufRef<'a, 'b> {
+    /// Creates a new `ReadBufRef` referencing the same `ReadBuf` as this one.
+    pub fn reborrow(&mut self) -> ReadBufRef<'_, 'b> {
+        ReadBufRef { read_buf: self.read_buf }
+    }
+
+    /// Returns a mutable reference to the filled portion of the buffer.
+    #[inline]
+    pub fn filled_mut(&mut self) -> &mut [u8] {
```

Comment:
I think rather than duplicating the API of ReadBuf, it would be better to Deref to it

Reply:
Since the point is to avoid allowing use of `mem::replace`, that would defeat the purpose, no?

Reply:
Yeah, true

Reply:
We can do things the other way around: keep only APIs on `ReadBufRef` and remove them from `ReadBuf`
- Cons: probably a little less ergonomic? (need to call `.borrow()` all the time)
- Pros: no API duplication